cpp/src/arrow/compute/exec/expression_test.cc

```diff
@@ -69,6 +70,39 @@ Expression true_unless_null(Expression argument) {
   return call("true_unless_null", {std::move(argument)});
 }
 
+Expression add(Expression l, Expression r) {
+  return call("add", {std::move(l), std::move(r)});
+}
+
+namespace arrow_literals {
+
+using namespace std::chrono_literals;
+
+inline auto operator""_ts(const char* c, size_t s) {
+  return [s = StringScalar(std::string{c, s})](TimeUnit::type unit) {
+    return literal(*s.CastTo(timestamp(unit)));
+  };
+}
+
+}  // namespace arrow_literals
+
+template <typename Rep, typename Period>
+Expression literal(std::chrono::duration<Rep, Period> d) {
+  int64_t int_value = d.count();
+  TimeUnit::type unit;
+  if constexpr (std::is_same_v<Period, std::nano>) {
+    unit = TimeUnit::NANO;
+  } else if constexpr (std::is_same_v<Period, std::micro>) {
+    unit = TimeUnit::MICRO;
+  } else if constexpr (std::is_same_v<Period, std::milli>) {
+    unit = TimeUnit::MILLI;
+  } else {
+    unit = TimeUnit::SECOND;
+    int_value = std::chrono::duration_cast<std::chrono::seconds>(d).count();
```

Comment:
So we grab the largest possible resolution and then rely on casting later?  Seems reasonable but just asking for clarification.

Reply:
I'll add a comment. If the argument is `std::chrono::nanoseconds`, then use TimeUnit::NANO. If the argument is `std::chrono::microseconds`, ... If the argument is none of `std::chrono::{nanoseconds, microseconds, milliseconds}` then use seconds